File: gem/lib/foaf/balances.rb
# frozen_string_literal: true

require "bigdecimal"

module Foaf
  module Balances
    module_function

    # Raw, creditor-oriented /trustlines row -> viewer perspective. The raw
    # balance is creditor-view, so the debtor sees it negated.
    def from_trustline_row(row)
      balance = decimal(fetch(row, "balance"))
      {
        viewer_balance: -balance,
        my_credit_limit: decimal(fetch(row, "received")),
        their_credit_limit: decimal(fetch(row, "given"))
      }
    end

    # Per-viewer userTrustlines row -> viewer perspective. These rows are ALREADY
    # viewer-oriented (the per-user endpoint applied the frame), so there is NO
    # negation: `balance` is already the viewer's balance (negative = the viewer
    # owes), `received` is the viewer's credit limit, `given` is the
    # counterparty's. Mirrors the TS `viewerBalance` for the read-side hooks.
    def from_user_trustline_row(row)
      {
        viewer_balance: decimal(fetch(row, "balance")),
        my_credit_limit: decimal(fetch(row, "received")),
        their_credit_limit: decimal(fetch(row, "given"))
      }
    end

    def fetch(row, key)
      return row[key] if row.respond_to?(:key?) && row.key?(key)
      symbol = key.to_sym
      return row[symbol] if row.respond_to?(:key?) && row.key?(symbol)

      raise KeyError, "missing FOAF trustline field #{key}"
    end
    private_class_method :fetch

    def decimal(value)
      BigDecimal(value.to_s)
    end
    private_class_method :decimal
  end
end

File: client/src/ledger/balances.ts
import type { TrustlineRow, ViewerBalance } from './types';

function decimal(value: unknown): string {
  if (typeof value === 'number' || typeof value === 'string') return String(value);
  return '0';
}

/**
 * Projects a per-viewer `userTrustlines` row into the viewer's perspective.
 *
 * These rows are ALREADY viewer-oriented: the network's per-user endpoint has
 * applied the viewer's frame before it hits the client. So there is NO sign or
 * limit flip here — `balance` is already the viewer's balance (negative = the
 * viewer owes, positive = the counterparty owes the viewer), `received` is the
 * viewer's credit limit, and `given` is the counterparty's. This mirrors the
 * gem's `Foaf::Balances.from_user_trustline_row` (also no negation).
 *
 * Contrast the RAW creditor-oriented `/trustlines` endpoint, where the balance
 * IS creditor-view and must be negated for the debtor — that conversion lives
 * in the gem's `from_trustline_row` (`viewer_balance = -balance`) and does not
 * apply here.
 *
 * The `viewerAddress`/`counterPartyAddress` params are kept for signature
 * stability (callers still identify the pair); they no longer drive a flip.
 * Wire keys (`given`/`received`) are primary; the write-side `creditline*`
 * names remain tolerant fallbacks. Strings are retained so applications do not
 * lose decimal precision.
 */
export function viewerBalance(
  row: TrustlineRow,
  _viewerAddress: string,
  _counterPartyAddress: string,
): ViewerBalance {
  const balance = decimal(row.balance);
  const given = decimal(row.given ?? row.creditlineGiven ?? row.creditline_given);
  const received = decimal(row.received ?? row.creditlineReceived ?? row.creditline_received);
  return {
    balance: balance === '-0' ? '0' : balance,
    creditlineGiven: given,
    creditlineReceived: received,
  };
}
